pyphare/pyphare/pharesee/geometry.py

```diff
@@ -0,0 +1,277 @@
+                        other_ovrlp= toFieldBox(other_ovrlp, ref_pd)
+
+
+                    overlaps.append({"pdatas":(ref_pd, cmp_pd),
+                                     "box" : other_ovrlp,
+                                     "offset":(offset, 0)})
+
+
+    return overlaps
+
+
+
+
+def hierarchy_overlaps(hierarchy):
+    """
+    returns all overlaps for the given hierarchy
+    """
+    overlaps = {}
+    for ilvl, lvl in enumerate(hierarchy.patch_levels):
+        overlaps[ilvl] = compute_overlaps(lvl.patches, hierarchy.refined_domain_box(ilvl))
+    return overlaps
+
+
+
+def get_periodic_list(patches, domain_box):
+    """
+    given a list of patches and a domain box the function
+    returns a list of patches sorted by origin where the
+    first (resp. last) patch is the last (resp. first) patch
+    of the sorted list, if that patch touches the upper (resp. lower)
+    border of the domain
+    """
+    from copy import copy
+
+    sorted_patches = sorted(patches, key=lambda p: p.origin)
+
+    if touch_domain_border(sorted_patches[-1].box, domain_box, "upper"):
+        last_patch = copy(sorted_patches[-1])
+        shift_patch(last_patch, -domain_box.upper)
+        sorted_patches.insert(0, last_patch)
+
+    if touch_domain_border(sorted_patches[0].box, domain_box, "lower"):
+        first_patch = copy(sorted_patches[0])
+        shift_patch(first_patch, domain_box.upper)
+        sorted_patches.append(first_patch)
+
+    return sorted_patches
+
+
+
+
+
+
+def particle_ghost_area_boxes(hierarchy):
+    """
+    this function returns boxes representing ghost cell boxes for all levels
+    a ghost cell box is a box containing cells of contiguous AMR index not
+    contained in the domain box.
+
+    if  - = domain cell and o = ghost cell
+
+    patchdata = o o o o o  - - - - - - - - - - o o o o o
+    boxes =     ^-------^                      ^-------^
+
+    return : {level_number : [{"pdata":patch_data1, "boxes":ghost_boxes},
+                              {"pdata":patch_data2, "boxes":ghost_boxes}, ...]}
+    """
+    gaboxes = {}
+
+    for ilvl, lvl in enumerate(hierarchy.patch_levels):
+        for patch in lvl.patches:
+
+            patch_data = patch.patch_datas["particles"]
+            gbox = patch_data.ghost_box
+            box = patch.box
+
+            if ilvl not in gaboxes:
+                gaboxes[ilvl] = []
+
+            gaboxes[ilvl] += [{"pdata": patch_data, "boxes": boxm.remove(gbox, box)}]
+
+    return gaboxes
+
+
+
+
+def level_ghost_boxes(hierarchy):
+    """
+    this function returns boxes representing level ghost cell boxes for all levels
+
+    A level ghost cell box is a ghost cell box that does not overlap any cell contained
+    in a patchData interior
+
+
+    patchdata1           : o o o o o - - - - - - - - - - - - o o o o o
+    patchdata2           :                               o o o o o - - - - - - - - -
+    lvl ghost cell boxes :                                   ^---^
+
```

Comment:
ok, I think I get it : this happens when the domain of this ghost box is not described at this level, but at a lower one... right ?

Reply:
I don't really understand what you mean by "this happens".
A level ghost cell is a cell of a patchdata that belongs to the ghost area (i.e. not in domain) and that does NOT overlap a domain cell of another patchdata in the same level.